### PR TITLE
Add Qwen3 Audio Encoder

### PR DESCRIPTION
**Pull request by eitanporat**
# Description

- Add audio encoder support for qwen3
* classes such as `AudioEncoder` and `Qwen3OmniAudioEncoder`
* added the `use_audio` flag to the configuration.
* various audio specific flags

- added merging of audio tokens for qwen3
- added support in maxengine and maxtext_utils
- added support in decode.py

# Tests

tests/check_qwen3_omni_audio_vs_reference.py

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/configs/models/qwen3-omni-30b-a3b.yml [hunk 1]
```diff
@@ -56,3 +56,24 @@ num_position_embeddings_for_vit: 2304
...
+attention_dropout_for_audio: 0.0
+n_window_for_audio: 50
+n_window_infer_for_audio: 400
+conv_chunksize_for_audio: 500
+num_conv_layers_for_audio: 3
+max_timescale_for_audio: 10000.0
+max_sample_len_for_audio: 10000
+
+freeze_audio_encoder_params: false
+freeze_vision_encoder_params: false
```
Comment (hengtaoguo) [inline src/MaxText/configs/models/qwen3-omni-30b-a3b.yml:79]: Can you remove these two flags from the model config? I want to control them in SFT config instead.

File: src/MaxText/layers/decoders.py [hunk 2]
```diff
@@ -700,6 +713,8 @@ def __call__(
       image_masks: None | jnp.ndarray = None,
       kv_caches: list[jax.Array] | None = None,
       attention_metadata=None,
+      audio_embeddings: None | jnp.ndarray = None,
+      audio_mask: None | jnp.ndarray = None,
```
Comment (hengtaoguo) [inline src/MaxText/layers/decoders.py:717]: Can we change the name to plural `audio_masks` to align with `image_masks`?

File: src/MaxText/layers/embeddings.py [hunk 3]
```diff
@@ -1431,3 +1460,52 @@ def __call__(self, num_frames: int, height: int, width: int) -> Array:
       interpolated = interpolated.astype(self.fprop_dtype)
 
     return interpolated
+  
+@dataclasses.dataclass(repr=False)
+class SinusoidsPositionEmbedding(nnx.Module):
```
Comment (hengtaoguo) [inline src/MaxText/layers/embeddings.py:1468]: Can we reuse `PositionalEmbedding` above? It also implements classic sinusoidal position embeddings. Are there any differences between them?

File: src/MaxText/layers/encoders.py [hunk 4]
```diff
@@ -66,4 +66,38 @@ def __call__(self, input_images, deterministic=False):
     if len(self.vision_encoder_layer) > 1:
       # vision embedder / projection layer, not frozen in most cases, trained / finetuned together with main model
       embeddings = self.vision_encoder_layer[1](config=cfg, mesh=mesh)(embeddings)
+
+
+class AudioEncoder(nn.Module):
```
Comment (hengtaoguo) [inline src/MaxText/layers/encoders.py:71]: Is it convenient for you to directly make it an NNX module? We are going to migrate VisionEncoder to NNX too.

File: src/MaxText/layers/qwen3.py [hunk 5]
```diff
@@ -1641,6 +1642,357 @@ def qwen3omni_visionprojector_as_linen(config: Config, mesh: Mesh) -> nn.Module:
...
+    Returns:
+        Projected output of shape (num_chunks, seq_len, output_dim_for_audio)
+    """
+    hidden_states = self.proj1(hidden_states)
+    hidden_states = jax.nn.gelu(hidden_states)
+    hidden_states = self.proj2(hidden_states)
+    return hidden_states
+
+
+class Qwen3OmniAudioModel(nnx.Module):
```
Comment (hengtaoguo) [inline src/MaxText/layers/qwen3.py:1933]: Hi why do we have an `AudioModel`? IIUC, in the AudioEncoder, you were calling `[qwen3.qwen3omni_audioencoder_as_linen, qwen3.qwen3omni_audioprojector_as_linen]`, just like we did for the vision encoder [encoder, projection].

File: src/MaxText/decode.py [hunk 6]
```diff
@@ -150,6 +150,8 @@ def main(argv: Sequence[str]) -> None:
           padded_tokens=tokens,
           images=processor_outputs.pixel_values if config.use_multimodal else None,
           image_masks=processor_outputs.pixel_mask if config.use_multimodal and "llama4" in config.model_name else None,
+          audio_values=processor_outputs.audio_values if config.use_multimodal and config.use_audio else None,
+          audio_masks=processor_outputs.audio_mask if config.use_multimodal and config.use_audio else None,
```
Comment (hengtaoguo) [inline src/MaxText/decode.py:154]: I wonder if we can just use `use_audio` flag to control audio info?

Upon request we may rename `use_multimodal` to `use_vision` shortly. So these two flags should be decoupled, WDYT?

File: src/MaxText/layers/qwen3.py (hunk 5 above)
Comment (eitanporat) [reply src/MaxText/layers/qwen3.py:1933]: yes it should just be Model

Comment (eitanporat) [reply src/MaxText/layers/qwen3.py:1933]: actually how would you control `freeze_vision_encoder_params` and `freeze_audio_encoder_params`
also deepstack is different

Comment (hengtaoguo) [reply src/MaxText/layers/qwen3.py:1933]: I mean this `AudioModel` was not used anywhere. You have been using its basic components `audio_encoder + audio_projector` (this is also our preferred way). So should we remove this AudioModel?

Usually we tend to freeze the encoder, and only tune the projector+decoder in SFT.
